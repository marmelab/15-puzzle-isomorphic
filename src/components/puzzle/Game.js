import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Grid from './Grid';

import { ShowWhenOnline } from '../detectOffline';
import withLoader from '../withLoader';

import Button from '../ui/Button';
import Switch from '../ui/Switch';
const LoaderButton = withLoader(Button);

import { suggestFactory } from '../../services/suggestMoveService';

const labels = {
    off: 'Hide',
    on: 'Show',
    title: 'Display the numbers',
};

export default class Game extends Component {
    static propTypes = {
        currentGrid: PropTypes.array.isRequired,
        isVictory: PropTypes.bool,
        onClickTile: PropTypes.func.isRequired,
        resolvedGrid: PropTypes.array.isRequired,
    };

    state = {
        loadingAdvice: false,
        showTileNumbers: false,
        suggestedTile: 0,
    };

    handleOnClickTile = tile => {
        this.setState({
            suggestedTile: 0,
        });
        this.props.onClickTile(tile);
    };

    handleOnToggle = toggleState => {
        this.setState({
            showTileNumbers: toggleState,
        });
    };

    requestSuggest = async () => {
        const { currentGrid, resolvedGrid } = this.props;
        this.setState({ loadingAdvice: true });

        try {
            const { Tile } = await suggestFactory()(currentGrid, resolvedGrid);
            this.setState({
                loadingAdvice: false,
                suggestedTile: Tile,
            });
        } catch (error) {
            console.error(error);
            // TODO : catch the errors in order to display them to the user.
        }
    };

    handleClickSuggest = () => {
        const { loadingAdvice } = this.state;
        if (loadingAdvice) {
            return;
        }
        this.requestSuggest();
    };

    render() {
        const { currentGrid, isVictory, resolvedGrid } = this.props;
        const { showTileNumbers, suggestedTile, loadingAdvice } = this.state;

        return (
            <div>
                <Grid
                    onClick={this.handleOnClickTile}
                    grid={currentGrid}
                    readOnly={isVictory}
                    resolvedGrid={resolvedGrid}
                    showNumbers={showTileNumbers}
                    tileToHighlight={suggestedTile}
                />
                <Switch labels={labels} onToggle={this.handleOnToggle} />
                {!isVictory && (
                    <div className="center">
                        <ShowWhenOnline>
                            <LoaderButton
                                isLoading={loadingAdvice}
                                size="small"
                                icon="help_outline"
                                label="Ask for help"
                                onClick={this.handleClickSuggest}
                            />
                        </ShowWhenOnline>
                    </div>
                )}
            </div>
        );
    }
}
