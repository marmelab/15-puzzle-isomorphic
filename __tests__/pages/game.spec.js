import React from 'react';
import renderer from 'react-test-renderer';

import Game, { title } from '../../pages/game';

// FIXME: this page is using a component dynamically loaded using the next/dynamic module. This bug will make the tests fail.
// You can find the related issue here : https://github.com/zeit/next.js/issues/3345
xdescribe('Game', () => {
    test('should render Game without error', () => {
        renderer.create(<Game />);
    });

    test('should render the Game correctly', () => {
        const component = renderer.create(<Game />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('title', () => {
    test('should render a turn', () => {
        const expectedTitle = '12 moves';
        expect(title(false, false, 12)).toEqual(expectedTitle);
    });

    test('should render the starting message', () => {
        const expectedTitle = 'Start the game by moving a tile';
        expect(title(false, false, 0)).toEqual(expectedTitle);
    });

    test('should render a victory', () => {
        const expectedTitle =
            'Congratulations, you have solved the puzzle in 12 turns!';
        expect(title(false, true, 12)).toEqual(expectedTitle);
    });

    test('should render a loading', () => {
        const expectedTitle = 'Building a new game';
        expect(title(true, false, 12)).toEqual(expectedTitle);
    });
});
