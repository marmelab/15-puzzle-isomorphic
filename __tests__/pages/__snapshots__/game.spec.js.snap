// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game should render the Game correctly 1`] = `
<div>
  <nav
    className="blue lighten-1"
    role="navigation"
  >
    <div
      className="nav-wrapper container"
    >
      <a
        className="brand-logo"
        href="/"
        onClick={[Function]}
      >
        15 Puzzle Isomorphic!
      </a>
    </div>
  </nav>
  <section
    className="container"
  >
    <div
      className="section"
    >
      <div
        className="bloc z-depth-3"
      >
        <div
          className="col s12"
        >
          <h5>
            Building a new game
          </h5>
        </div>
        <div
          className="col s12"
        >
          <div
            className="activity-indicator-wrapper"
          >
            <div
              className="preloader-wrapper big active"
            >
              <div
                className="spinner-layer spinner-blue"
              >
                <div
                  className="circle-clipper left"
                >
                  <div
                    className="circle"
                  />
                </div>
                <div
                  className="gap-patch"
                >
                  <div
                    className="circle"
                  />
                </div>
                <div
                  className="circle-clipper right"
                >
                  <div
                    className="circle"
                  />
                </div>
              </div>
              <div
                className="spinner-layer spinner-red"
              >
                <div
                  className="circle-clipper left"
                >
                  <div
                    className="circle"
                  />
                </div>
                <div
                  className="gap-patch"
                >
                  <div
                    className="circle"
                  />
                </div>
                <div
                  className="circle-clipper right"
                >
                  <div
                    className="circle"
                  />
                </div>
              </div>
              <div
                className="spinner-layer spinner-yellow"
              >
                <div
                  className="circle-clipper left"
                >
                  <div
                    className="circle"
                  />
                </div>
                <div
                  className="gap-patch"
                >
                  <div
                    className="circle"
                  />
                </div>
                <div
                  className="circle-clipper right"
                >
                  <div
                    className="circle"
                  />
                </div>
              </div>
              <div
                className="spinner-layer spinner-green"
              >
                <div
                  className="circle-clipper left"
                >
                  <div
                    className="circle"
                  />
                </div>
                <div
                  className="gap-patch"
                >
                  <div
                    className="circle"
                  />
                </div>
                <div
                  className="circle-clipper right"
                >
                  <div
                    className="circle"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="section"
    >
      <div
        className="row"
      >
        <div
          className="col s12"
        >
          <div
            className="buttons-wrapper"
          >
            <a
              className="btn red"
              href="/"
              onClick={[Function]}
            >
              <i
                className="material-icons left"
              >
                keyboard_return
              </i>
              Back to home
            </a>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
